engine/crates/engine-v2/src/request/walkers/operation_limits.rs

```diff
@@ -0,0 +1,86 @@
+use std::collections::HashSet;
+
+use schema::StringId;
+
+use super::{BoundSelectionSetWalker, BoundSelectionWalker};
+
+impl<'a> BoundSelectionSetWalker<'a, ()> {
+    pub(crate) fn max_depth(&self) -> u16 {
+        (*self)
+            .into_iter()
+            .map(|selection| match selection {
+                BoundSelectionWalker::Field(field) => {
+                    field
+                        .selection_set()
+                        .map(|selection_set| selection_set.max_depth())
+                        .unwrap_or_default()
+                        + 1
+                }
+                BoundSelectionWalker::InlineFragment(inline) => inline.selection_set().max_depth(),
+                BoundSelectionWalker::FragmentSpread(spread) => spread.selection_set().max_depth(),
+            })
+            .max()
+            .expect("must be defined")
+    }
+
+    pub(crate) fn alias_count(&self) -> u16 {
+        (*self)
+            .into_iter()
+            .map(|selection| match selection {
+                BoundSelectionWalker::Field(field) => {
+                    (field.definition().as_field().expect("must be a field").name() == field.response_key_str()) as u16
```

Comment:
`as_field()` returns `None` for `__typename` fields. Actually, you should add the following to the `BoundAnyFieldDefinitionWalker` (`walkers/field_definition`):

```
    pub fn schema_name(&self) -> &'a str {
        match self.as_ref() {
            BoundAnyFieldDefinition::TypeName(_) => "__typename",
            BoundAnyFieldDefinition::Field(definition) => self.schema_walker.walk(definition.field_id).name(),
        }
    }
```

and compare that against the `response_key_str()` like you do here.

Reply:
Thanks 🙂